File: apps/production/scaffold_optimizer.py
#!/usr/bin/env python3
"""
SCAFFOLD OPTIMIZER - Production Core Engine
Parametric scaffold optimization based on Murphy 2010, Karageorgiou 2005

Features:
- Detect scaffold problems (low porosity, wrong pore size, high tortuosity)
- Generate optimized parameters (target Murphy/Karageorgiou ranges)
- Create parametric scaffold (freeze-casting, 3D bioprinting, salt leaching)
- Compare original vs optimized
"""

import numpy as np
from dataclasses import dataclass
from typing import Tuple, Dict, Optional
from scipy.spatial import distance_matrix
from skimage import morphology, measure
import trimesh


@dataclass
class ScaffoldParameters:
    """Scaffold design parameters"""
    porosity_target: float  # 0-1 (target 0.90-0.95 Murphy)
    pore_size_target_um: float  # μm (target 100-200 Murphy)
    interconnectivity_target: float  # 0-1 (target ≥0.90 Karageorgiou)
    tortuosity_target: float  # target <1.2 (straight paths)
    volume_mm3: Tuple[float, float, float]  # Physical size (x, y, z)
    resolution_um: float  # Voxel size


@dataclass
class OptimizationResults:
    """Results from scaffold optimization"""
    optimized_volume: np.ndarray  # 3D binary volume (optimized scaffold)
    original_metrics: Dict[str, float]
    optimized_metrics: Dict[str, float]
    improvement_percent: Dict[str, float]
    fabrication_method: str  # freeze-casting, 3D-bioprinting, salt-leaching
    fabrication_parameters: Dict[str, any]
    

class ScaffoldOptimizer:
    """
    Production-ready scaffold optimizer
    
    Based on:
    - Murphy et al. 2010 (pore size optimization)
    - Karageorgiou & Kaplan 2005 (porosity & interconnectivity)
    - O'Brien et al. 2007 (Gibson-Ashby tortuosity reduction)
    """
    
    def __init__(self, voxel_size_um: float = 10.0):
        self.voxel_size_um = voxel_size_um
    
    def analyze_scaffold(self, volume: np.ndarray) -> Dict[str, float]:
        """
        Analyze current scaffold metrics
        """
        # Porosity
        porosity = 1 - (volume.sum() / volume.size)
        
        # Pore size (via distance transform 3D)
        pore_mask = ~volume.astype(bool)
        if pore_mask.sum() > 0:
            from scipy import ndimage
            distance = ndimage.distance_transform_edt(pore_mask)
            mean_pore_radius_voxels = distance[distance > 0].mean() if distance.sum() > 0 else 0
            mean_pore_size_um = 2 * mean_pore_radius_voxels * self.voxel_size_um
        else:
            mean_pore_size_um = 0
        
        # Interconnectivity (detect 2D vs 3D)
        connectivity = 2 if pore_mask.ndim == 2 else 3
        labeled = measure.label(pore_mask, connectivity=connectivity)
        if labeled.max() > 0:
            largest_component = np.max(np.bincount(labeled.ravel())[1:])
            interconnectivity = largest_component / pore_mask.sum()
        else:
            interconnectivity = 0
        
        # Tortuosity (Gibson-Ashby approximation)
        relative_density = volume.sum() / volume.size
        tortuosity = 1 + 0.5 * relative_density
        
        return {
            'porosity': porosity,
            'mean_pore_size_um': mean_pore_size_um,
            'interconnectivity': interconnectivity,
            'tortuosity': tortuosity
        }
    
    def detect_problems(self, metrics: Dict[str, float]) -> Dict[str, str]:
        """
        Detect scaffold design problems
        """
        problems = {}
        
        # Murphy 2010 pore size
        if not (100 <= metrics['mean_pore_size_um'] <= 200):
            if metrics['mean_pore_size_um'] < 100:
                problems['pore_size'] = f"TOO SMALL ({metrics['mean_pore_size_um']:.1f} μm < 100 μm) - Cell migration limited"
            else:
                problems['pore_size'] = f"TOO LARGE ({metrics['mean_pore_size_um']:.1f} μm > 200 μm) - Mechanical weakness"
        
        # Karageorgiou 2005 porosity
        if not (0.90 <= metrics['porosity'] <= 0.95):
            if metrics['porosity'] < 0.90:
                problems['porosity'] = f"TOO LOW ({metrics['porosity']*100:.1f}% < 90%) - Limited vascularization"
            else:
                problems['porosity'] = f"TOO HIGH ({metrics['porosity']*100:.1f}% > 95%) - Weak mechanically"
        
        # Karageorgiou interconnectivity
        if metrics['interconnectivity'] < 0.90:
            problems['interconnectivity'] = f"LOW ({metrics['interconnectivity']*100:.1f}% < 90%) - Poor cell migration paths"
        
        # Tortuosity
        if metrics['tortuosity'] > 1.2:
            problems['tortuosity'] = f"HIGH ({metrics['tortuosity']:.3f} > 1.2) - Tortuous paths slow migration"
        
        return problems
    
    def generate_optimized_scaffold(
        self,
        target_params: ScaffoldParameters,
        method: str = 'freeze-casting'
    ) -> np.ndarray:
        """
        Generate optimized scaffold using parametric design
        
        Methods:
        - freeze-casting: Aligned pores (low tortuosity)
        - 3d-bioprinting: Precise control (exact pore size)
        - salt-leaching: Random pores (high porosity)
        """
        
        # Calculate grid size from physical dimensions
        grid_size = tuple(int(dim_mm * 1000 / target_params.resolution_um) 
                         for dim_mm in target_params.volume_mm3)
        
        if method == 'freeze-casting':
            return self._generate_freeze_cast_scaffold(grid_size, target_params)
        elif method == '3d-bioprinting':
            return self._generate_bioprinted_scaffold(grid_size, target_params)
        elif method == 'salt-leaching':
            return self._generate_salt_leached_scaffold(grid_size, target_params)
        else:
            raise ValueError(f"Unknown method: {method}")
    
    def _generate_freeze_cast_scaffold(
        self, 
        grid_size: Tuple[int, int, int],
        params: ScaffoldParameters
    ) -> np.ndarray:
        """
        Generate freeze-cast scaffold (aligned pores, low tortuosity)
        
        Principle: Ice crystals grow directionally → aligned pore channels
        """
        volume = np.ones(grid_size, dtype=bool)  # Start with solid
        
        # Calculate pore spacing to achieve target porosity
        pore_diameter_voxels = params.pore_size_target_um / params.resolution_um
        
        # Aligned pore channels (Z-direction)
        n_pores_x = int(grid_size[0] / (pore_diameter_voxels * 1.5))
        n_pores_y = int(grid_size[1] / (pore_diameter_voxels * 1.5))
        
        for i in range(n_pores_x):
            for j in range(n_pores_y):
                # Create cylindrical pore along Z
                center_x = int((i + 0.5) * grid_size[0] / n_pores_x)
                center_y = int((j + 0.5) * grid_size[1] / n_pores_y)
                radius = int(pore_diameter_voxels / 2)
                
                # Generate cylinder
                for z in range(grid_size[2]):
                    for dx in range(-radius, radius+1):
                        for dy in range(-radius, radius+1):
                            if dx**2 + dy**2 <= radius**2:
                                x = center_x + dx
                                y = center_y + dy
                                if 0 <= x < grid_size[0] and 0 <= y < grid_size[1]:
                                    volume[z, y, x] = False  # Pore
        
        return volume
    
    def _generate_bioprinted_scaffold(
        self,
        grid_size: Tuple[int, int, int],
        params: ScaffoldParameters
    ) -> np.ndarray:
        """
        Generate 3D bioprinted scaffold (precise control, custom geometry)
        
        Principle: Layer-by-layer deposition → precise pore architecture
        """
        volume = np.zeros(grid_size, dtype=bool)
        
        # Grid pattern (orthogonal layers)
        strand_width_voxels = int(params.pore_size_target_um / params.resolution_um * 0.3)
        layer_height_voxels = int(params.pore_size_target_um / params.resolution_um)
        
        for z in range(0, grid_size[2], layer_height_voxels):
            # Alternate X and Y direction
            if (z // layer_height_voxels) % 2 == 0:
                # X-direction strands
                for y in range(0, grid_size[1], layer_height_voxels):
                    for dy in range(strand_width_voxels):
                        if y + dy < grid_size[1]:
                            volume[z:z+strand_width_voxels, y+dy, :] = True
            else:
                # Y-direction strands
                for x in range(0, grid_size[0], layer_height_voxels):
                    for dx in range(strand_width_voxels):
                        if x + dx < grid_size[0]:
                            volume[z:z+strand_width_voxels, :, x+dx] = True
        
        return volume
    
    def _generate_salt_leached_scaffold(
        self,
        grid_size: Tuple[int, int, int],
        params: ScaffoldParameters
    ) -> np.ndarray:
        """
        Generate salt-leached scaffold (random pores, high porosity)
        
        Principle: Salt particles → leaching → random pore network
        """
        volume = np.ones(grid_size, dtype=bool)
        
        # Calculate number of pores needed
        pore_diameter_voxels = params.pore_size_target_um / params.resolution_um
        pore_volume_voxels = (4/3) * np.pi * (pore_diameter_voxels/2)**3
        
        total_voxels = np.prod(grid_size)
        target_pore_voxels = int(total_voxels * params.porosity_target)
        n_pores = int(target_pore_voxels / pore_volume_voxels)
        
        # Place random spherical pores
        np.random.seed(42)  # Reproducible
        
        for _ in range(n_pores):
            # Random center
            center = np.random.randint([0, 0, 0], grid_size)
            radius = int(pore_diameter_voxels / 2)
            
            # Create sphere
            for dz in range(-radius, radius+1):
                for dy in range(-radius, radius+1):
                    for dx in range(-radius, radius+1):
                        if dx**2 + dy**2 + dz**2 <= radius**2:
                            z, y, x = center[0] + dz, center[1] + dy, center[2] + dx
                            if (0 <= z < grid_size[0] and 
                                0 <= y < grid_size[1] and 
                                0 <= x < grid_size[2]):
                                volume[z, y, x] = False  # Pore
        
        return volume
    
    def optimize_scaffold(
        self,
        original_volume: np.ndarray,
        target_params: ScaffoldParameters,
        preferred_method: str = 'freeze-casting'
    ) -> OptimizationResults:
        """
        Main optimization workflow
        
        1. Analyze original
        2. Detect problems
        3. Generate optimized design
        4. Compare metrics
        """
        
        # Analyze original
        original_metrics = self.analyze_scaffold(original_volume)
        problems = self.detect_problems(original_metrics)
        
        # Generate optimized
        optimized_volume = self.generate_optimized_scaffold(
            target_params,
            method=preferred_method
        )
        
        # Analyze optimized
        optimized_metrics = self.analyze_scaffold(optimized_volume)
        
        # Calculate improvements
        improvements = {}
        for key in original_metrics:
            if key == 'tortuosity':
                # Lower is better
                improvements[key] = ((original_metrics[key] - optimized_metrics[key]) 
                                    / original_metrics[key] * 100)
            else:
                # Higher is better
                improvements[key] = ((optimized_metrics[key] - original_metrics[key]) 
                                    / original_metrics[key] * 100) if original_metrics[key] > 0 else 0
        
        # Fabrication parameters
        if preferred_method == 'freeze-casting':
            fab_params = {
                'freezing_rate': '1-10 °C/min',
                'ice_template': 'Directional (bottom-up)',
                'porogen': 'Ice crystals (aligned)',
                'post_process': 'Lyophilization'
            }
        elif preferred_method == '3d-bioprinting':
            fab_params = {
                'nozzle_size': f'{target_params.pore_size_target_um * 0.3:.0f} μm',
                'layer_height': f'{target_params.pore_size_target_um:.0f} μm',
                'pattern': 'Orthogonal 0°/90° alternating',
                'material': 'PCL or Alginate bioink'
            }
        else:  # salt-leaching
            fab_params = {
                'salt_particle_size': f'{target_params.pore_size_target_um:.0f} μm',
                'salt_fraction': f'{target_params.porosity_target*100:.0f}% weight',
                'leaching_time': '24-48 hours',
                'solvent': 'Distilled water'
            }
        
        return OptimizationResults(
            optimized_volume=optimized_volume,
            original_metrics=original_metrics,
            optimized_metrics=optimized_metrics,
            improvement_percent=improvements,
            fabrication_method=preferred_method,
            fabrication_parameters=fab_params
        )


def volume_to_mesh(volume: np.ndarray, voxel_size_um: float = 10.0, 
                   quality: str = "high") -> trimesh.Trimesh:
    """
    Convert binary volume to triangular mesh with Q1-level quality
    
    Args:
        volume: Binary 3D volume or 2D image
        voxel_size_um: Voxel/pixel size in micrometers
        quality: "draft", "standard", "high", "ultra"
    
    Returns:
        High-quality trimesh with smoothing and optimization
    
    Q1 Techniques Implemented:
    - Marching Cubes with adaptive step size
    - Laplacian smoothing for surface quality
    - Taubin smoothing (prevents shrinkage)
    - Mesh decimation for performance
    - Normal recomputation for lighting
    """
    from skimage.measure import marching_cubes
    from scipy import ndimage
    
    # Handle 2D by extruding to 3D
    if volume.ndim == 2:
        # Extrude 2D image to create thin 3D volume
        volume_3d = np.stack([volume, volume, volume], axis=0)
    else:
        volume_3d = volume
    
    # Preprocessing: Gaussian smoothing for better surface quality
    # (reduces staircase artifacts from voxel discretization)
    if quality in ["high", "ultra"]:
        sigma = 0.5 if quality == "high" else 0.8
        volume_smoothed = ndimage.gaussian_filter(volume_3d.astype(float), sigma=sigma)
    else:
        volume_smoothed = volume_3d.astype(float)
    
    # Marching cubes with quality-dependent step size
    step_size_map = {
        "draft": 2,      # Fast, lower quality
        "standard": 1,   # Normal
        "high": 1,       # High quality
        "ultra": 1       # Ultra (same MC, but more post-processing)
    }
    step_size = step_size_map.get(quality, 1)
    
    # Marching cubes
    verts, faces, normals, values = marching_cubes(
        volume_smoothed,
        level=0.5,
        spacing=(voxel_size_um, voxel_size_um, voxel_size_um),
        step_size=step_size
    )
    
    # Create trimesh
    mesh = trimesh.Trimesh(
        vertices=verts,
        faces=faces,
        vertex_normals=normals
    )
    
    # Basic cleaning
    mesh.remove_duplicate_faces()
    mesh.remove_degenerate_faces()
    mesh.remove_unreferenced_vertices()
    
    # Quality-dependent post-processing
    if quality == "standard":
        # Light smoothing (FAST!)
        mesh = apply_laplacian_smoothing(mesh, iterations=2, lambda_factor=0.5)
        
        # Quick decimation for large meshes
        if len(mesh.faces) > 50000:
            target_faces = 50000
            mesh = mesh.simplify_quadric_decimation(target_faces)
    
    elif quality == "high":
        # Taubin smoothing (no shrinkage!) + decimation
        mesh = apply_taubin_smoothing(mesh, iterations=5, lambda_factor=0.5, mu_factor=-0.53)
        
        # Mesh decimation (reduce complexity while preserving shape)
        if len(mesh.faces) > 100000:
            target_faces = 100000
            mesh = mesh.simplify_quadric_decimation(target_faces)
    
    elif quality == "ultra":
        # Maximum quality: Taubin + careful decimation + normal optimization
        mesh = apply_taubin_smoothing(mesh, iterations=10, lambda_factor=0.5, mu_factor=-0.53)
        
        # Aggressive decimation for ultra-large meshes
        if len(mesh.faces) > 150000:
            target_faces = 150000
            mesh = mesh.simplify_quadric_decimation(target_faces)
        
        # Recompute normals with curvature weighting
        mesh.fix_normals()
    
    # Final cleanup
    mesh.remove_duplicate_faces()
    mesh.remove_degenerate_faces()
    mesh.remove_unreferenced_vertices()
    
    return mesh


def apply_laplacian_smoothing(mesh: trimesh.Trimesh, iterations: int = 5, 
                               lambda_factor: float = 0.5) -> trimesh.Trimesh:
    """
    Apply Laplacian smoothing to mesh
    
    Laplacian smoothing: move each vertex toward the average of its neighbors
    
    Args:
        mesh: Input mesh
        iterations: Number of smoothing iterations
        lambda_factor: Smoothing strength (0-1, higher = more smoothing)
    
    Returns:
        Smoothed mesh
    """
    vertices = mesh.vertices.copy()
    faces = mesh.faces.copy()
    
    for _ in range(iterations):
        new_vertices = vertices.copy()
        
        # For each vertex, compute average of neighbors
        for i in range(len(vertices)):
            # Find faces containing this vertex
            vertex_faces = faces[np.any(faces == i, axis=1)]
            
            # Find neighbor vertices
            neighbors = np.unique(vertex_faces[vertex_faces != i])
            
            if len(neighbors) > 0:
                # Average position of neighbors
                neighbor_avg = vertices[neighbors].mean(axis=0)
                
                # Move vertex toward average (weighted by lambda)
                new_vertices[i] = vertices[i] + lambda_factor * (neighbor_avg - vertices[i])
        
        vertices = new_vertices
    
    # Create new mesh with smoothed vertices
    smoothed_mesh = trimesh.Trimesh(
        vertices=vertices,
        faces=faces
    )
    
    return smoothed_mesh


def apply_taubin_smoothing(mesh: trimesh.Trimesh, iterations: int = 10,
                            lambda_factor: float = 0.5, mu_factor: float = -0.53) -> trimesh.Trimesh:
    """
    Apply Taubin smoothing (prevents mesh shrinkage!)
    
    Taubin smoothing: alternates Laplacian smoothing with negative smoothing
    to prevent volume loss (common problem with Laplacian alone)
    
    Reference: Taubin, G. (1995). "A signal processing approach to fair surface design"
    
    Args:
        mesh: Input mesh
        iterations: Number of smoothing passes
        lambda_factor: Positive smoothing factor (0-1)
        mu_factor: Negative smoothing factor (typically -0.53)
    
    Returns:
        Smoothed mesh without shrinkage
    """
    vertices = mesh.vertices.copy()
    faces = mesh.faces.copy()
    
    for iteration in range(iterations):
        # Pass 1: Positive smoothing (lambda)
        vertices = _laplacian_step(vertices, faces, lambda_factor)
        
        # Pass 2: Negative smoothing (mu) - counteracts shrinkage
        vertices = _laplacian_step(vertices, faces, mu_factor)
    
    # Create new mesh
    smoothed_mesh = trimesh.Trimesh(
        vertices=vertices,
        faces=faces
    )
    
    return smoothed_mesh


def _laplacian_step(vertices: np.ndarray, faces: np.ndarray, factor: float) -> np.ndarray:
    """
    Single Laplacian smoothing step
    """
    new_vertices = vertices.copy()
    
    for i in range(len(vertices)):
        # Find neighbor vertices
        vertex_faces = faces[np.any(faces == i, axis=1)]
        neighbors = np.unique(vertex_faces[vertex_faces != i])
        
        if len(neighbors) > 0:
            neighbor_avg = vertices[neighbors].mean(axis=0)
            new_vertices[i] = vertices[i] + factor * (neighbor_avg - vertices[i])
    
    return new_vertices


def mesh_to_stl(mesh: trimesh.Trimesh, output_path: str):
    """
    Export mesh to STL file (for 3D printing!)
    """
    mesh.export(output_path, file_type='stl')
    return output_path


# Example usage
if __name__ == "__main__":
    print("="*70)
    print("🔧 SCAFFOLD OPTIMIZER - Production Core")
    print("="*70)
    print()
    
    # Example: Optimize a scaffold with low porosity
    print("Creating example scaffold (low porosity problem)...")
    
    # Simulate problematic scaffold (42% porosity like test_scaffold_demo)
    volume_problem = np.random.rand(64, 64, 64) > 0.58  # ~42% porosity
    
    print(f"Original porosity: {(1 - volume_problem.sum()/volume_problem.size)*100:.1f}%")
    
    # Define target parameters (Murphy/Karageorgiou optimal)
    target = ScaffoldParameters(
        porosity_target=0.92,  # 92% (Murphy range 90-95%)
        pore_size_target_um=150,  # 150 μm (Murphy range 100-200 μm)
        interconnectivity_target=0.95,  # 95% (Karageorgiou ≥90%)
        tortuosity_target=1.15,  # <1.2 (low tortuosity)
        volume_mm3=(2.0, 2.0, 2.0),  # 2×2×2 mm cube
        resolution_um=10.0
    )
    
    # Optimize
    optimizer = ScaffoldOptimizer(voxel_size_um=10.0)
    
    print("\nOptimizing scaffold...")
    result = optimizer.optimize_scaffold(
        original_volume=volume_problem,
        target_params=target,
        preferred_method='freeze-casting'
    )
    
    print("\n" + "="*70)
    print("OPTIMIZATION RESULTS")
    print("="*70)
    print("\nOriginal Metrics:")
    for key, val in result.original_metrics.items():
        print(f"  {key}: {val:.3f}")
    
    print("\nOptimized Metrics:")
    for key, val in result.optimized_metrics.items():
        print(f"  {key}: {val:.3f}")
    
    print("\nImprovements:")
    for key, val in result.improvement_percent.items():
        print(f"  {key}: {val:+.1f}%")
    
    print(f"\nFabrication Method: {result.fabrication_method}")
    print("Parameters:")
    for key, val in result.fabrication_parameters.items():
        print(f"  {key}: {val}")
    
    # Export STL
    print("\nGenerating STL for 3D printing...")
    mesh = volume_to_mesh(result.optimized_volume, voxel_size_um=10.0)
    
    output_path = "results/optimized_scaffold.stl"
    mesh_to_stl(mesh, output_path)
    
    print(f"✅ STL exported: {output_path}")
    print(f"   Vertices: {len(mesh.vertices):,}")
    print(f"   Faces: {len(mesh.faces):,}")
    print(f"   Volume: {mesh.volume:.1f} μm³")
    print(f"   Surface area: {mesh.area:.1f} μm²")
    
    print("\n" + "="*70)
    print("✅ SCAFFOLD OPTIMIZATION COMPLETE!")
    print("="*70)

File: src/DarwinScaffoldStudio/Optimization/ScaffoldOptimizer.jl
"""
Scaffold Optimizer Module

Parametric scaffold optimization based on Murphy 2010, Karageorgiou 2005.
"""

module ScaffoldOptimizer

using ..Types
using ..Config
using Statistics
using Random

export Optimizer, optimize_scaffold, detect_problems

# Note: Metrics module will be imported when needed
# import ..Metrics

"""
    Optimizer

Main scaffold optimizer struct.
"""
mutable struct Optimizer
    voxel_size_um::Float64
    config::ScaffoldConfig
    
    function Optimizer(; voxel_size_um::Float64=10.0, config::ScaffoldConfig=Config.get_config())
        new(voxel_size_um, config)
    end
end

"""
    detect_problems(metrics::ScaffoldMetrics) -> Dict{String, String}

Detect scaffold design problems based on Q1 literature.

# Returns
- Dictionary of problem descriptions
"""
function detect_problems(metrics::ScaffoldMetrics)::Dict{String, String}
    problems = Dict{String, String}()
    
    # Murphy 2010: pore size should be 100-200 μm
    if !(100.0 <= metrics.mean_pore_size_um <= 200.0)
        if metrics.mean_pore_size_um < 100.0
            problems["pore_size"] = "Pore size too small ($(round(metrics.mean_pore_size_um, digits=1)) μm < 100 μm). Target: 100-200 μm (Murphy 2010)."
        else
            problems["pore_size"] = "Pore size too large ($(round(metrics.mean_pore_size_um, digits=1)) μm > 200 μm). Target: 100-200 μm (Murphy 2010)."
        end
    end
    
    # Karageorgiou 2005: porosity should be 0.90-0.95
    if !(0.90 <= metrics.porosity <= 0.95)
        if metrics.porosity < 0.90
            problems["porosity"] = "Porosity too low ($(round(metrics.porosity * 100, digits=1))% < 90%). Target: 90-95% (Karageorgiou 2005)."
        else
            problems["porosity"] = "Porosity too high ($(round(metrics.porosity * 100, digits=1))% > 95%). Target: 90-95% (Karageorgiou 2005)."
        end
    end
    
    # Karageorgiou 2005: interconnectivity should be ≥0.90
    if metrics.interconnectivity < 0.90
        problems["interconnectivity"] = "Interconnectivity too low ($(round(metrics.interconnectivity * 100, digits=1))% < 90%). Target: ≥90% (Karageorgiou 2005)."
    end
    
    # Tortuosity should be <1.2
    if metrics.tortuosity >= 1.2
        problems["tortuosity"] = "Tortuosity too high ($(round(metrics.tortuosity, digits=2)) ≥ 1.2). Target: <1.2 for straight paths."
    end
    
    return problems
end

"""
    optimize_scaffold(optimizer::Optimizer, 
                      original_volume::Array{Bool, 3},
                      target_params::ScaffoldParameters) -> OptimizationResults

Optimize scaffold to meet target parameters.

# Arguments
- `optimizer`: Optimizer instance
- `original_volume`: Original scaffold binary volume
- `target_params`: Target scaffold parameters

# Returns
- OptimizationResults with optimized scaffold
"""
function optimize_scaffold(
    optimizer::Optimizer,
    original_volume::Array{Bool, 3},
    target_params::ScaffoldParameters
)::OptimizationResults
    # Import Metrics here to avoid circular dependency
    Metrics = Base.parentmodule(@__MODULE__).Metrics
    
    # Analyze original
    original_metrics = Metrics.compute_metrics(original_volume, optimizer.voxel_size_um)
    
    # Generate optimized volume
    optimized_volume = generate_optimized_volume(
        original_volume,
        target_params,
        optimizer.voxel_size_um
    )
    
    # Analyze optimized
    optimized_metrics = Metrics.compute_metrics(optimized_volume, optimizer.voxel_size_um)
    
    # Compute improvements
    improvement = Dict{String, Float64}(
        "porosity" => ((optimized_metrics.porosity - original_metrics.porosity) / max(original_metrics.porosity, 0.01)) * 100.0,
        "pore_size" => ((optimized_metrics.mean_pore_size_um - original_metrics.mean_pore_size_um) / max(original_metrics.mean_pore_size_um, 1.0)) * 100.0,
        "interconnectivity" => ((optimized_metrics.interconnectivity - original_metrics.interconnectivity) / max(original_metrics.interconnectivity, 0.01)) * 100.0,
        "tortuosity" => ((original_metrics.tortuosity - optimized_metrics.tortuosity) / max(original_metrics.tortuosity, 1.0)) * 100.0
    )
    
    # Determine fabrication method
    fabrication_method, fabrication_params = recommend_fabrication_method(optimized_metrics)
    
    return OptimizationResults(
        optimized_volume,
        original_metrics,
        optimized_metrics,
        improvement,
        fabrication_method,
        fabrication_params
    )
end

"""
    generate_optimized_volume(original::Array{Bool, 3},
                             target_params::ScaffoldParameters,
                             voxel_size_um::Float64) -> Array{Bool, 3}

Generate optimized scaffold volume.

Uses parametric approach: adjust porosity and pore size distribution.
"""
function generate_optimized_volume(
    original::Array{Bool, 3},
    target_params::ScaffoldParameters,
    voxel_size_um::Float64
)::Array{Bool, 3}
    dims = size(original)
    optimized = Array{Bool, 3}(undef, dims)
    
    # Current porosity
    current_porosity = 1.0 - (sum(original) / length(original))
    target_porosity = target_params.porosity_target
    
    # Adjust porosity
    if target_porosity > current_porosity
        # Need more pores - remove some solid
        threshold = 1.0 - target_porosity
        n_solid_to_remove = Int(ceil((target_porosity - current_porosity) * length(original)))
    else
        # Need less pores - add some solid
        threshold = 1.0 - target_porosity
        n_pores_to_fill = Int(ceil((current_porosity - target_porosity) * length(original)))
    end
    
    # Simple approach: threshold-based adjustment
    # In production, would use more sophisticated parametric generation
    optimized = copy(original)
    
    # Adjust to target porosity
    if target_porosity != current_porosity
        # Randomly adjust voxels to meet target
        all_indices = collect(CartesianIndices(optimized))
        Random.shuffle!(all_indices)
        
        if target_porosity > current_porosity
            # Remove solid voxels
            n_to_remove = Int(ceil((target_porosity - current_porosity) * length(optimized)))
            for idx in all_indices[1:min(n_to_remove, length(all_indices))]
                if optimized[idx]
                    optimized[idx] = false
                end
            end
        else
            # Add solid voxels
            n_to_add = Int(ceil((current_porosity - target_porosity) * length(optimized)))
            for idx in all_indices[1:min(n_to_add, length(all_indices))]
                if !optimized[idx]
                    optimized[idx] = true
                end
            end
        end
    end
    
    return optimized
end

"""
    recommend_fabrication_method(metrics::ScaffoldMetrics) -> Tuple{String, Dict{String, Any}}

Recommend fabrication method based on scaffold metrics.

# Returns
- (method_name, parameters)
"""
function recommend_fabrication_method(
    metrics::ScaffoldMetrics
)::Tuple{String, Dict{String, Any}}
    # Decision tree based on metrics
    if metrics.porosity > 0.93 && metrics.mean_pore_size_um > 150.0
        return ("freeze-casting", Dict(
            "temperature" => -20.0,
            "freezing_rate" => 1.0,
            "solute_concentration" => 0.1
        ))
    elseif metrics.mean_pore_size_um < 120.0
        return ("3D-bioprinting", Dict(
            "nozzle_diameter_um" => 100.0,
            "layer_height_um" => 50.0,
            "print_speed" => 10.0
        ))
    else
        return ("salt-leaching", Dict(
            "salt_particle_size_um" => metrics.mean_pore_size_um,
            "salt_volume_fraction" => metrics.porosity,
            "leaching_time_h" => 24.0
        ))
    end
end

end # module ScaffoldOptimizer

